docs/dg/dev/architecture/coding-conventions.md

```diff
@@ -0,0 +1,2165 @@
+</details>
+
+**Guideline**
+- `Actions` should not contain any logic that is outside the regular responsibilities of a `Controller` (see description above).
+- `Controller` has an inherited `castId()` method that should be used for casting numerical IDs.
+- The inherited `getFactory()` method grants access to the [Factory](#factory) layer.
+- The inherited `getFacade()` or `getClient()` methods grant access to the corresponding [facade](#facade-design-pattern) functionalities.
+
+**Example**
+```php
+<?php
+
+namespace Pyz\Zed\ConfigurableBundleGui\Communication\Controller;
+
+/**
+ * @method \Pyz\Zed\ConfigurableBundleGui\Communication\ConfigurableBundleGuiCommunicationFactory getFactory()
+ * @method \Spryker\Zed\ConfigurableBundleGui\Business\ConfigurableBundleGuiFacadeInterface getFacade()
+ */
+class TemplateController extends Spryker\Zed\ConfigurableBundleGui\Communication\Controller\TemplateController
+{
+    /**
+     * @var string
+     */
+    protected const PARAM_ID_CONFIGURABLE_BUNDLE_TEMPLATE = 'id-configurable-bundle-template';
+    
+    /**
+     * @param \Symfony\Component\HttpFoundation\Request $request
+     *
+     * @return \Symfony\Component\HttpFoundation\JsonResponse
+     */
+    public function slotTableAction(Request $request): JsonResponse
+    {
+        $idConfigurableBundleTemplate = $this->castId(
+            $request->get(static::PARAM_ID_CONFIGURABLE_BUNDLE_TEMPLATE),
+        );
+
+        $table = $this->getFactory()->createConfigurableBundleTemplateSlotTable($idConfigurableBundleTemplate);
+
+        return $this->jsonResponse($table->fetchData());
+    }
+}
+```
+
+## Dependency Provider
+
+```
+[Organization]
+├── Zed
+│   └── [Module]        
+│       ├── [Module]Config.php
+│       └── [Module]DependencyProvider.php
+├── Yves
+│   └── [Module]        
+│       ├── [Module]Config.php
+│       └── [Module]DependencyProvider.php
+├── Glue
+│   └── [Module]        
+│       ├── [Module]Config.php
+│       └── [Module]DependencyProvider.php
+├── Client
+│   └── [Module]        
+│       ├── [Module]Config.php
+│       └── [Module]DependencyProvider.php
+└── Service
+    └── [Module]        
+        ├── [Module]Config.php
+        └── [Module]DependencyProvider.php
+```
+
+**Description**
+
+Injects required dependencies to a module application layer. Typically, dependencies are [facades](#facade-design-pattern) or [plugins](#plugin). 
+
+Dependency injection is orchestrated through a `provide-add-get` structure (see example below).
+- The `provide` method is the highest level that holds only `add` calls without any additional logic.
+- The `add` method is the middle level that injects the dependencies into the dependency container using a class constant and a late-binding instantiating closure.
+- The `get` method is the lowest level that sources the dependency.
+
+**Conventions**
+- All class constants MUST be `public` (to decrease conflicts in definition for being a public API class).
```

Comment:
Regardless of conflicts, they have to be public. Because, all injected dependencies are expected to be used in Factory classes. The class constant will be needed in Factory class.

```suggestion
- All class constants MUST be `public`.
```